### PR TITLE
feat: reports hash via flashbots

### PR DESCRIPTION
**Pull request by b0dhidharma**

### CODE REVIEW

File: packages/stealth-txs/scripts/watcher/yearn.ts [hunk 1]
```diff
@@ -35,125 +49,218 @@ const generateRandomNumber = (min: number, max: number): string => {
...
 };
 
+async function setup(): Promise<void> {
+  console.log(`[Setup] Starting on network ${hardhatArguments.network!}(${chainId}) ...`);
+  alive.startCheck();
+  console.log('[Setup] Getting reporter ...');
+  [reporterSigner] = await ethers.getSigners();
+  web3ReporterSigner = web3.eth.accounts.privateKeyToAccount(
+    await kms.decrypt((chainId === 1 ? process.env.MAINNET_1_PRIVATE_KEY : process.env.GOERLI_1_PRIVATE_KEY) as string)
+  );
```
Comment (b0dhidharma) [inline packages/stealth-txs/scripts/watcher/yearn.ts:59]: Before I was decrypting private keys in each iteration, didn't make any sense.

File: packages/stealth-txs/scripts/watcher/yearn.ts [hunk 2]
```diff
@@ -35,125 +49,218 @@ const generateRandomNumber = (min: number, max: number): string => {
...
     });
     stealthVault.on('PenaltyApplied', (hash: string, caller: string, penalty: BigNumber, reporter: string) => {
+      console.log('[Main] Event PenaltyApplied received');
       reduceBond(caller, penalty);
       addBond(reporter, penalty.div(10));
     });
   });
 }
 
 async function checkTx(tx: Transaction) {
```
Comment (b0dhidharma) [inline packages/stealth-txs/scripts/watcher/yearn.ts:156]: Whole analysis tx is around ~7 ms.

File: packages/stealth-txs/scripts/watcher/yearn.ts [hunk 3]
```diff
@@ -35,125 +49,218 @@ const generateRandomNumber = (min: number, max: number): string => {
...
+      submitted = true;
+    } else if (resolution == FlashbotsBundleResolution.BlockPassedWithoutInclusion) {
+      console.log('[Flashbot] BlockPassedWithoutInclusion, re-build and re-send bundle');
+    } else if (resolution == FlashbotsBundleResolution.AccountNonceTooHigh) {
+      console.log('[Flashbot] AccountNonceTooHigh, adjust nonce');
+      rejected = true;
+    }
+    targetBlock += 1;
+  }
+  return submitted;
```
Comment (b0dhidharma) [inline packages/stealth-txs/scripts/watcher/yearn.ts:247]: I really don't like this. But this is just a script, not an app. We could take sometime to scale this into a dapp that takes care of every case / side-case. Handles nonces and bundles in a really smart way, but thats not even close to an MVP.

File: packages/commons/utils/commons.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { Transaction } from 'ethers';
+import { encode } from 'rlp';
+
+export const normalizeAddress = (address: string): string => address.toLowerCase();
+
+// Ref: https://eips.ethereum.org/EIPS/eip-2718, https://eips.ethereum.org/EIPS/eip-2930 and https://eips.ethereum.org/EIPS/eip-1559
+export const getRawTransaction = (transaction: Transaction): string => {
```
Comment (b0dhidharma) [inline packages/commons/utils/commons.ts:7]: This took some time to get right. I was able to test it with type == 2 and type == 0, was not able to test with type == 1.

Comment (skeletor-spaceman) [reply packages/commons/utils/commons.ts:7]: oh my, nice job

**Review by skeletor-spaceman — APPROVED** [review]
Awesome job!

**Pull request merged**